source/server/http/admin.cc

```diff
@@ -661,15 +662,19 @@ envoy::admin::v2alpha::ServerInfo::State AdminImpl::serverState() {
 
 Http::Code AdminImpl::handlerServerInfo(absl::string_view, Http::HeaderMap& headers,
                                         Buffer::Instance& response, AdminStream&) {
-  time_t current_time = time(nullptr);
+  const auto current_time = server_.timeSource().systemTime();
   envoy::admin::v2alpha::ServerInfo server_info;
   server_info.set_version(VersionInfo::version());
   server_info.set_state(serverState());
 
-  server_info.mutable_uptime_current_epoch()->set_seconds(current_time -
-                                                          server_.startTimeCurrentEpoch());
-  server_info.mutable_uptime_all_epochs()->set_seconds(current_time -
-                                                       server_.startTimeFirstEpoch());
+  const auto current_uptime = std::chrono::duration_cast<std::chrono::seconds>(
```

Comment:
since it's not lexically obvious, can you given an explicit type here? Maybe uint64_t?

Also you are substracting system-time which is not necessarily monotonic, so this could be negative. Do you want to use MonotonicTime as a type and timeSource().monotonicTime() to get the time?

Reply:
alternatively I'm not sure if we need the time to be printable, in which case you should keep using system-time, but test for a>b before subtracting b-a. I'm not sure what guarantees are made when you print monotonic time.

Reply:
The docs for `std::chrono::steady_clock` suggest it is "most suitable for measuring intervals." so sounds like that might be the right match here

https://en.cppreference.com/w/cpp/chrono/steady_clock